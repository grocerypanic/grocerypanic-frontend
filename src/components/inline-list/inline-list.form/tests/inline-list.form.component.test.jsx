import { fireEvent, render, screen } from "@testing-library/react";
import React from "react";
import ItemListForm, { testIDs } from "../inline-list.form.component";

jest.mock("../../../../configuration/theme", () => {
  return {
    __esModule: true,
    ...jest.requireActual("../../../../configuration/theme"),
    ui: { alertTimeout: 10 },
  };
});

jest.mock("../../../../configuration/theme");

const initialProps = {
  item: { id: -1, name: "mockItem" },
  handleSave: jest.fn(),
  setErrorMsg: jest.fn(),
  setSelected: jest.fn(),
  transaction: false,
};

describe("InlineList", () => {
  let currentProps;
  let container;

  beforeEach(() => {
    jest.clearAllMocks();
    resetProps();
  });

  const resetProps = () => {
    currentProps = { ...initialProps };
  };

  const arrange = () => {
    const result = render(<ItemListForm {...currentProps} />);
    container = result.container;
  };

  describe("when outside of a transaction", () => {
    let form;
    let button;

    beforeEach(async () => {
      arrange();
      form = await screen.findByTestId(testIDs.ListItemNewItemInputElement);
      button = await screen.findByTestId(testIDs.ListItemSaveButton);
    });

    it("should display the entry form with the correct props", () => {
      expect(form).toHaveFocus();
      expect(form).toHaveProperty("id", "newItem");
      expect(form).toHaveProperty("type", "text");
      expect(form).toHaveProperty("name", "newItem");
      expect(form.required).toBeFalsy();
      expect(form.readOnly).toBeFalsy();
      expect(form.defaultValue).toBe(currentProps.item.name);
    });

    it("should display a save button with correct props", () => {
      expect(button).toHaveStyle({ height: "40px" });
    });

    it("should match the snapshot on file (styles)", () => {
      expect(container).toMatchSnapshot();
    });

    describe("when the form is clicked", () => {
      beforeEach(() => {
        expect(currentProps.setSelected).toBeCalledTimes(0);
        fireEvent.click(form);
      });

      it("should set the new item to selected", () => {
        expect(currentProps.setSelected).toBeCalledTimes(1);
        expect(currentProps.setSelected).toBeCalledWith(-1);
      });
    });

    describe("when the form is changed", () => {
      beforeEach(() => {
        expect(currentProps.setSelected).toBeCalledTimes(0);
        fireEvent.change(form, {
          target: { value: "new value" },
        });
      });

      it("should clear any error message", () => {
        expect(currentProps.setErrorMsg).toBeCalledTimes(1);
        expect(currentProps.setErrorMsg).toBeCalledWith(null);
      });
    });

    describe("when the form is changed", () => {
      let testValue = "new value";

      beforeEach(() => {
        expect(currentProps.setSelected).toBeCalledTimes(0);
        fireEvent.change(form, {
          target: { value: testValue },
        });
      });

      it("should clear any error message", () => {
        expect(currentProps.setErrorMsg).toBeCalledTimes(1);
        expect(currentProps.setErrorMsg).toBeCalledWith(null);
      });

      describe("when the save button is clicked", () => {
        beforeEach(() => {
          expect(currentProps.handleSave).toBeCalledTimes(0);
          fireEvent.click(button);
        });

        it("should call handleSave with the form value", () => {
          expect(currentProps.handleSave).toBeCalledTimes(1);
          expect(currentProps.handleSave).toBeCalledWith(testValue);
        });
      });
    });
  });

  describe("when a transaction is ongoing", () => {
    let form;

    beforeEach(async () => {
      currentProps.transaction = true;
      arrange();
      form = await screen.findByTestId(testIDs.ListItemNewItemInputElement);
    });

    it("should display a READONLY entry form with the correct props", () => {
      expect(form).toHaveFocus();
      expect(form).toHaveProperty("id", "newItem");
      expect(form).toHaveProperty("type", "text");
      expect(form).toHaveProperty("name", "newItem");
      expect(form.required).toBeFalsy();
      expect(form.readOnly).toBeTruthy();
      expect(form.defaultValue).toBe(currentProps.item.name);
    });

    it("should NOT display a save button", () => {
      expect(screen.queryByTestId(testIDs.ListItemSaveButton)).toBeNull();
    });

    it("should match the snapshot on file (styles)", () => {
      expect(container).toMatchSnapshot();
    });

    describe("when the form is clicked", () => {
      beforeEach(() => {
        expect(currentProps.setSelected).toBeCalledTimes(0);
        fireEvent.click(form);
      });

      it("should NOT set the new item to selected", () => {
        expect(currentProps.setSelected).toBeCalledTimes(0);
      });
    });
  });
});
