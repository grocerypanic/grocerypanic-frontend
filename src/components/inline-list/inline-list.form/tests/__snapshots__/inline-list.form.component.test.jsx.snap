// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InlineList when a transaction is ongoing should match the snapshot on file (styles) 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: left;
  -webkit-box-align: left;
  -ms-flex-align: left;
  align-items: left;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  background: #FFFFFF;
  margin-bottom: 5px;
  padding: 5px;
  color: blue;
  min-width: 220px;
}

.c0 div:last-child {
  margin-left: auto;
}

.c0 .inline-list-item-title-div:hover {
  color: red;
  cursor: pointer;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: left;
  -webkit-box-align: left;
  -ms-flex-align: left;
  align-items: left;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 40px;
}

<div>
  <form>
    <li
      class="c0"
      data-testid="ListItemElement"
    >
      <div
        class="c1"
      >
        <input
          data-testid="ListItemNewItemInputElement"
          id="newItem"
          name="newItem"
          readonly=""
          required=""
          size="15"
          type="text"
          value="mockItem"
        />
      </div>
    </li>
  </form>
</div>
`;

exports[`InlineList when outside of a transaction should match the snapshot on file (styles) 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: left;
  -webkit-box-align: left;
  -ms-flex-align: left;
  align-items: left;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  background: #FFFFFF;
  margin-bottom: 5px;
  padding: 5px;
  color: blue;
  min-width: 220px;
}

.c0 div:last-child {
  margin-left: auto;
}

.c0 .inline-list-item-title-div:hover {
  color: red;
  cursor: pointer;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: left;
  -webkit-box-align: left;
  -ms-flex-align: left;
  align-items: left;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 40px;
}

<div>
  <form>
    <li
      class="c0"
      data-testid="ListItemElement"
    >
      <div
        class="c1"
      >
        <input
          data-testid="ListItemNewItemInputElement"
          id="newItem"
          name="newItem"
          required=""
          size="15"
          type="text"
          value="mockItem"
        />
      </div>
      <div>
        <button
          class="btn btn-success"
          data-testid="ListItemSaveButton"
          style="height: 40px;"
        >
          <span>
            save
          </span>
        </button>
      </div>
    </li>
  </form>
</div>
`;
